ipaserver/install/ipa_idrange_fix.py

```diff
@@ -0,0 +1,1073 @@
+    id_entities = []
+    if user:
+        id_name = "user"
+        ldap_filter = get_outofrange_filter(id_ranges, "account", "uid")
+    else:
+        id_name = "group"
+        ldap_filter = get_outofrange_filter(id_ranges, "group", "gid")
+
+    logger.debug("Searching %ss in %s with filter: %s", id_name, container_dn,
+                 ldap_filter)
+    try:
+        identities = api.Backend.ldap2.get_entries(
+            container_dn,
+            ldap.SCOPE_ONELEVEL,
+            ldap_filter,
+        )
+        for entry in identities:
+            id_entities.append(read_identity(entry, user))
+    except errors.NotFound:
+        logger.debug("No out of range %ss found in %s!", id_name, container_dn)
+    except errors.ExecutionError as e:
+        logger.error("Exception while reading %s: %s", container_dn, e)
+    return id_entities
+
+
+def read_identity(ldapentry, user: bool = True) -> IDentity:
+    """Function to convert LDAP entry to IDentity object"""
+    sv = ldapentry.single_value
+    id_entity = IDentity()
+    id_entity.dn = ldapentry.dn
+    id_entity.name = sv.get("cn")
+    id_entity.number = (
+        int(sv.get("uidNumber")) if user else int(sv.get("gidNumber"))
+    )
+    id_entity.user = user
+    logger.debug("Out of range found: %s", id_entity.debug())
+    return id_entity
+
+
+def get_outofrange_filter(
+    id_ranges_all: list[IDRange], object_class: str, attr: str
+) -> str:
+    """Function to create LDAP filter for out of range users and groups"""
+    # we need to look only for ipa-local ranges
+    id_ranges = get_ipa_local_ranges(id_ranges_all)
+
+    ldap_filter = f"(&(objectClass=posix{object_class})(|"
+
+    # adding gaps in ranges to the filter
+    for i in range(len(id_ranges) + 1):
+        if i == 0:
+            start_condition = f"({attr}Number>=1)"
+        else:
+            start_condition = f"({attr}Number>={id_ranges[i - 1].last_id + 1})"
+
+        if i < len(id_ranges):
+            end_condition = f"({attr}Number<={id_ranges[i].first_id - 1})"
+        else:
+            end_condition = f"({attr}Number<=2147483647)"
+
+        ldap_filter += f"(&{start_condition}{end_condition})"
+
+    ldap_filter += "))"
+
+    return ldap_filter
+# endregion
+# Writing to LDAP
+# region
+
+
+def apply_ridbases(id_range: IDRange) -> None:
+    """Funtion to apply RID bases to the range in LDAP"""
+    try:
+        api.Backend.ldap2.modify_s(
+            id_range.dn,
+            [
+                (ldap.MOD_ADD, "ipaBaseRID", str(id_range.base_rid)),
+                (
+                    ldap.MOD_ADD,
+                    "ipaSecondaryBaseRID",
+                    str(id_range.secondary_base_rid),
+                ),
+            ],
+        )
+        logger.info("RID bases updated for range '%s'", id_range.name)
+
+    except ldap.CONSTRAINT_VIOLATION as e:
+        logger.error(
+            "Failed to add RID bases to the range '%s'.",
+            id_range.name
+        )
```

Comment:
This loses the _msg part of the constraing violation. That could be useful to determine why it failed.

Reply:
fixed. Not sure how I missed it here